File: src/T3.Web/ClientApp/src/app/set/models/set-commit-body.ts
import {Score} from "./score";

interface SetCommitBodyBase<T extends string> {
  type: T;
}

export interface SetCommitBodyNoOp extends SetCommitBodyBase<'NoOp'> {
  type: 'NoOp';
}

export interface SetCommitBodySetScoreChange extends SetCommitBodyBase<'SetScoreChange'> {
  type: 'SetScoreChange';
  setScoreDelta: Score;
}

export type SetCommitBody = SetCommitBodyNoOp
  | SetCommitBodySetScoreChange

File: src/T3.Web/Services/Set/Models/SetCommitBody.cs
namespace T3.Web.Services.Set.Models;
public enum SetCommitBodyType
{
    Invalid,
    NoOp,
    SetScoreChange,
}

public abstract class SetCommitBody
{
    public virtual SetCommitBodyType Type { get; }
}

public class SetCommitBodyNoOp : SetCommitBody
{
    public override SetCommitBodyType Type => SetCommitBodyType.NoOp;
}

public class SetCommitBodySetScoreChange : SetCommitBody
{
    public override SetCommitBodyType Type => SetCommitBodyType.SetScoreChange;
    public Score SetScoreDelta { get; set; }
}